packages/@lwc/engine-core/src/framework/freeze-template.ts

```diff
@@ -0,0 +1,130 @@
+            return ArrayFill;
+        case 'splice':
+            return ArraySplice;
+        case 'copyWithin':
+            return ArrayCopyWithin;
+    }
+}
+
+let mutationWarningsSilenced = false;
+
+// Warn if the user tries to mutate tmpl.stylesheets, e.g.:
+// `tmpl.stylesheets.push(someStylesheetFunction)`
+function warnOnArrayMutation(stylesheets: TemplateStylesheetFactories) {
+    // We can't handle users calling Array.prototype.slice.call(tmpl.stylesheets), but
+    // we can at least warn when they use the most common mutation methods.
+    for (const prop of ARRAY_MUTATION_METHODS) {
+        const originalArrayMethod = getOriginalArrayMethod(prop);
+        stylesheets[prop] = function arrayMutationWarningWrapper() {
+            logError(
+                `Mutating the "stylesheets" array on a template function ` +
+                    `is deprecated and may be removed in a future version of LWC.`
+            );
+            // @ts-ignore
+            return originalArrayMethod.apply(this, arguments);
+        };
+    }
+}
+
+// TODO [#2782]: eventually freezeTemplate() will _actually_ freeze the tmpl object. Today it
+// just warns on mutation.
+export function freezeTemplate(tmpl: Template) {
+    if (process.env.NODE_ENV !== 'production') {
+        if (!isUndefined(tmpl.stylesheets)) {
+            warnOnArrayMutation(tmpl.stylesheets);
+        }
+        for (const prop of TEMPLATE_PROPS) {
+            let value = tmpl[prop];
+            defineProperty(tmpl, prop, {
+                enumerable: true,
+                configurable: true,
+                get() {
+                    return value;
+                },
+                set(newValue) {
+                    if (!mutationWarningsSilenced) {
+                        logError(
+                            `Dynamically setting the "${prop}" property on a template function ` +
+                                `is deprecated and may be removed in a future version of LWC.`
+                        );
+                    }
+                    value = newValue;
+                },
+            });
+        }
+
+        const originalDescriptor = getOwnPropertyDescriptor(tmpl, 'stylesheetTokens');
+        defineProperty(tmpl, 'stylesheetTokens', {
+            enumerable: true,
+            configurable: true,
+            get: originalDescriptor!.get,
+            set(value) {
+                if (process.env.NODE_ENV !== 'production') {
+                    logError(
+                        `Dynamically setting the "stylesheetTokens" property on a template function ` +
+                            `is deprecated and may be removed in a future version of LWC.`
+                    );
+                }
+                // Avoid logging twice (for both stylesheetToken and stylesheetTokens)
+                mutationWarningsSilenced = true;
+                originalDescriptor!.set!.call(this, value);
+                mutationWarningsSilenced = false;
```

Comment:
Do we want a warning to show in the console for each `set`, or a single warning when it is done the first time?

Reply:
Good question. I considered doing the bookkeeping to ensure we only warn once (per component that calls the setter), but since I don't expect these setters to be called very often, I thought the risk of the console being flooded was low, and wasn't worth the extra code for the bookkeeping.